std-bits/table/src/main/java/org/enso/table/parsing/TypeInferringParser.java

```diff
@@ -0,0 +1,55 @@
+package org.enso.table.parsing;
+
+import org.enso.table.data.column.builder.object.Builder;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.column.storage.StringStorage;
+import org.enso.table.parsing.problems.ProblemAggregator;
+import org.enso.table.read.WithProblems;
+
+/**
+ * The type inferring parser tries to parse the given column using a set of provided parsers. It
+ * returns the result of the first parser that succeeds without reporting any problems.
+ *
+ * <p>At least one parser must always be provided. The last parser on the list is used as a fallback
+ * - its result is returned regardless of whether it contained problems or not.
+ */
+public class TypeInferringParser implements DatatypeParser {
+
+  private final IncrementalDatatypeParser[] baseParsers;
+
+  public TypeInferringParser(IncrementalDatatypeParser[] baseParsers) {
+    if (baseParsers.length == 0) {
+      throw new IllegalArgumentException("At least one parser must be provided.");
+    }
+    this.baseParsers = baseParsers;
+  }
+
+  @Override
+  public WithProblems<Storage> parseColumn(StringStorage sourceStorage) {
+    parsers:
+    for (int i = 0; i < baseParsers.length; ++i) {
+      boolean isLast = i == baseParsers.length - 1;
+
+      IncrementalDatatypeParser parser = baseParsers[i];
+      Builder builder = parser.makeBuilderWithCapacity(sourceStorage.size());
+      var aggregator = new ProblemAggregator();
+
+      for (int j = 0; j < sourceStorage.size(); ++j) {
+        String cell = sourceStorage.getItem(j);
+        if (cell != null) {
+          Object parsed = parser.parseSingleValue(cell, aggregator);
+          if (aggregator.hasProblems() && !isLast) {
+            continue parsers;
+          }
+          builder.appendNoGrow(parsed);
+        } else {
+          builder.appendNoGrow(null);
+        }
+      }
+
+      return new WithProblems<>(builder.seal(), aggregator.getAggregatedProblems());
+    }
+
+    throw new IllegalStateException("`baseParsers` should not be empty.");
```

Comment:
This is an incorrect final message (constructor verifies this).

Should just return original storage unchanged:
```java
return new WithProblems<>(sourceStorage, new ArrayList<>());
```

Reply:
I don't agree.

This place should _never_ be reached. I wanted to mark this (can still do) with `assert false`, but since without `-ea` assertions are not checked, Java would still complain about a missing return statement. Alternatively I can:
```
assert false;
return null;
```

to make Java happy. I decided that throwing an exception is a neat middle-ground. I could change the message to "Internal error: this code should never be reached", or to the above if you think that's better. But I don't think returning unchanged is the way to go as it can hide bugs.

Reply:
As mentioned above I think we should remove the IdentityParser which would then result in getting here. I think it is a cleaner approach.

Reply:
Hmm, I see. In the conjunction with the comment above that does make sense. However, in the CSV parsing mode, I will always have a fallback pass - the one that removes quotes and parses quoted text as regular text. So there will be some copying anyway. And this fallback pass needs to be able to report warnings (mismatched quote), so I cannot let it then fall back to the non-parsing pure identity one.